### PR TITLE
[Issue 119] Publishing snapshot artifacts to JCenter

### PR DESCRIPTION
**Pull request by vijikarthi**
**Change log description**

This PR addresses the issue https://github.com/pravega/flink-connectors/issues/119. The master build artifacts of the connector project will be published to JCenter OJO snapshot repository. The publishing part will be handled in Jarvis.

* Introduced publishing option to OJO (snapshot repository)
* Introduced publishing option to Bintray (release repository). This is functional barring GPG signing but for the moment the plan will be to continue to use Maven central for release artifacts
* The snapshot artifacts are labeled in such a way that they can be uniquely identified. Hence any projects (samples repo) that depends on the connector (master) could draw these published artifacts instead of taking a source dependency (submodule)

**Purpose of the change**
* To support projects (nautilus-samples, pravega-samples etc.,) that requires the latest snapshot artifacts 

**What the code does**
Provides gradle configurations to publish artifacts to JCenter repository

**How to verify it**
> To publish to snapshot repo
* This step is automated through the Travis integration. Whenever a branch is merged in master, the successful completion of the build will trigger snapshot publish
* To manually try publishing to snapshot repo 
`./gradlew clean assemble publishToRepo -PpublishUrl=jcenterSnapshot -PpublishUsername=<user>  -PpublishPassword=<password>`

Virtual Repo: https://oss.jfrog.org/artifactory/jfrog-dependencies/io/pravega/pravega-connectors-flink_2.11/

> To publish to release repo (Maven Central)

`./gradlew clean assemble publishToRepo -PdoSigning=true -Psigning.password=<signing-password> -PpublishUrl=mavenCentral -PpublishUsername=<sonatype-username> -PpublishPassword=<sonatype-password>`

> To publish to Bintray (release repo but not yet ready for usage)

`./gradlew clean publishToBintray -PpublishRepo=bintrayRelease -PpublishUsername=<user> -PpublishPassword=<password>`

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -25,7 +28,19 @@ cache:
   - .gradle
   - $HOME/.gradle
   - $HOME/.m2
-  
+
+stages:
+  - name: build
+  - name: snapshot
+    if: branch = issue-119
```
Comment (vijikarthi) [inline .travis.yml]: I will change this to `master` before merging the code. It is pointing to the branch for testing purpose.

**Comment by vijikarthi** [comment]
FYI - I haven't tested the artifacts signing part for the snapshot publishing since I don't have the password. After we test the signing part,  we need to include the encrypted password in Travis configuration.

File: gradle/maven.gradle [hunk 2]
```diff
@@ -34,6 +34,17 @@ uploadShadow {
                         authentication(userName: publishUsername, password: publishPassword)
                     }
                 }
+                else if (publishUrl == "jcenterSnapshot") {
```
[the remark this answers is not in the document]
Comment (vijikarthi) [reply gradle/maven.gradle]: I agree that it does not actually represent the URL. I believe the variable `publishUrl` was introduced to pass the Artifactory (ASD) URL and later it was re-purposed to be used as a flag to indicate publishing to `mavenCentral`. I think we can leave the variable as-is or introduce new variable to indicate the repository type.

File: gradle/maven.gradle [hunk 3]
```diff
@@ -34,6 +34,17 @@ uploadShadow {
...
+                    repository(url: "https://oss.jfrog.org/artifactory/oss-snapshot-local/") {
+                        def bintrayUser = System.getProperty('BINTRAY_USER')
+                        def bintrayKey = System.getProperty('BINTRAY_KEY')
```
[the remark this answers is not in the document]
Comment (vijikarthi) [reply gradle/maven.gradle]: It is used to support jarvis deployment where the bintray credentials (userid/password) is provided as environment variable

File: build.gradle [hunk 4]
```diff
@@ -138,3 +140,12 @@ javadoc {
     failOnError = false
     exclude "**/impl/**";
 }
+
+def getProjectVersion() {
+    def stdout = new ByteArrayOutputStream()
+    exec {
+        commandLine "./version.sh", project
```
[the remark this answers is not in the document]
Comment (vijikarthi) [reply build.gradle]: replaced script with gradle-git plugin

File: gradle.properties [hunk 5]
```diff
@@ -17,9 +17,18 @@ mockitoVersion=1.10.19
 connectorVersion=0.3.0-SNAPSHOT
 pravegaVersion=0.3.0-SNAPSHOT
 apacheCommonsVersion=3.7
+artifactsBaseName=pravega-connectors-flink_2.11
```
[the remark this answers is not in the document]
Comment (vijikarthi) [reply gradle.properties]: It is parameterized since It is needed in the Bintray publishing (pom configuration)

File: gradle/bintray.gradle [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    }
+    pkg {
+        repo = "${bintrayRepoName}"
+        name = "${bintrayPackageName}"
+        userOrg = "${bintrayOrgName}"
+        licenses = ['Apache-2.0']
+        publish = "${autoPublish}"
+        vcsUrl = 'https://github.com/pravega/flink-connectors.git'
+        version {
+            desc = 'Flink Pravega Connector Artifacts'
```
[the remark this answers is not in the document]
Comment (vijikarthi) [reply gradle/bintray.gradle]: I believe there is no standard way to do this. I have seen some artifacts that are using project homepage URL as identifier. It is an optional field and I have added standard test for now. We can revisit this when we plan to use Bintray (release) publishing later?

**Comment by vijikarthi** [comment]
@EronWright I have made the changes and pushed the commit. Please take a look for another round of review.

**Comment by EronWright** [comment]
Thanks will do

File: gradle.properties (hunk 5 above)
[the remark this answers is not in the document]
Comment (EronWright) [reply gradle.properties]: I take your answer to mean that the string `pravega-connectors-flink_2.11` was needed in two places, and so you made an variable in `gradle.properties` as though it were a constant.    My issue is that values in `gradle.properties` are considered parameters and may be overridden.   Seems wholly unnecessary and stinky.

I think you may simply use `archivesBaseName` in your POM configuration block:
```
bintrayPublication(MavenPublication) { publication ->
   artifactId project.archivesBaseName

```

File: gradle/maven.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (EronWright) [reply gradle/maven.gradle]: OK but let's clean it up when we finalize the move to jcenter.

File: gradle/maven.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (EronWright) [reply gradle/maven.gradle]: Let me rephrase my concern.   The script accepts a _system property_ of `BINTRAY_USER` and an _environment variable_ of `BINTRAY_USER`.   System properties are not Gradle friendly.  Please accept only an environment variable.

[the remark this answers is not in the document]
Comment (vijikarthi) [reply gradle/maven.gradle]: I have moved the configuration to use gradle property instead of system property. Only for Travis environment, we need the environment variable.

**Comment by fpj** [comment]
@vijikarthi did you mean to say Travis here:

> The publishing part will be handled in Jarvis

File: build.gradle [hunk 7]
```diff
@@ -140,3 +144,15 @@ javadoc {
...
     exclude "**/impl/**";
 }
+
+def getProjectVersion() {
+    String ver = connectorVersion
+    if (ver.contains("-SNAPSHOT")) {
+        String versionLabel = ver.substring(0, ver.indexOf("-SNAPSHOT"))
+        def count = grgit.log(includes:['HEAD']).size()
+        def commitId = "${grgit.head().abbreviatedId}"
+        int curTime = new Date().getTime()/1000
```
[the remark this answers is not in the document]
Comment (vijikarthi) [reply build.gradle]: It is convenient to have the timestamp to make the build version unique if we have a need to manually test the publishing step multiple times for the same commit id.

[the remark this answers is not in the document]
Comment (EronWright) [reply build.gradle]: I am a bit skeptical of including the timestamp, because a) our goal is to link to specific commits, not to 'dirty' builds and b) jcenter assumedly keeps numerous builds for the same snapshot version.
